File: apps/intel-worker/src/lib/storage.ts
// Object storage behind one interface: GCS in prod, local FS in tests.
// Keys are the canonical paths from @agentronics/intel-schema (no gs:// prefix).

import { mkdir, readFile, writeFile, access } from "node:fs/promises";
import { dirname, join, normalize, sep } from "node:path";

export interface ObjectStorage {
  put(key: string, data: Buffer): Promise<void>;
  get(key: string): Promise<Buffer>;
  exists(key: string): Promise<boolean>;
  /** Fully-qualified URI for job bookkeeping (gs://… or file://…). */
  uri(key: string): string;
}

export class LocalFsStorage implements ObjectStorage {
  private readonly rootDir: string;

  constructor(rootDir: string) {
    this.rootDir = rootDir;
  }

  private resolve(key: string): string {
    const path = normalize(join(this.rootDir, key));
    if (!path.startsWith(normalize(this.rootDir) + sep)) {
      throw new Error(`key escapes storage root: ${key}`);
    }
    return path;
  }

  async put(key: string, data: Buffer): Promise<void> {
    const path = this.resolve(key);
    await mkdir(dirname(path), { recursive: true });
    await writeFile(path, data);
  }

  async get(key: string): Promise<Buffer> {
    return readFile(this.resolve(key));
  }

  async exists(key: string): Promise<boolean> {
    try {
      await access(this.resolve(key));
      return true;
    } catch {
      return false;
    }
  }

  uri(key: string): string {
    return `file://${this.resolve(key)}`;
  }
}

export class GcsStorage implements ObjectStorage {
  // Lazy import keeps @google-cloud/storage out of unit tests.
  private bucketPromise: Promise<import("@google-cloud/storage").Bucket> | undefined;
  private readonly bucketName: string;

  constructor(bucketName: string) {
    this.bucketName = bucketName;
  }

  private async bucket() {
    this.bucketPromise ??= import("@google-cloud/storage").then(
      ({ Storage }) => new Storage().bucket(this.bucketName)
    );
    return this.bucketPromise;
  }

  async put(key: string, data: Buffer): Promise<void> {
    const bucket = await this.bucket();
    await bucket.file(key).save(data, { resumable: false });
  }

  async get(key: string): Promise<Buffer> {
    const bucket = await this.bucket();
    const [data] = await bucket.file(key).download();
    return data;
  }

  async exists(key: string): Promise<boolean> {
    const bucket = await this.bucket();
    const [ok] = await bucket.file(key).exists();
    return ok;
  }

  uri(key: string): string {
    return `gs://${this.bucketName}/${key}`;
  }
}

export function storageFromEnv(): ObjectStorage {
  const bucket = process.env.GCS_BUCKET;
  if (!bucket) throw new Error("GCS_BUCKET is not set");
  return new GcsStorage(bucket);
}

File: apps/intel-ml/src/intel_ml/storage.py
"""Object storage behind one interface: GCS in prod, local FS in tests.

Keys are the canonical paths from packages/intel-schema (no gs:// prefix).
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from pathlib import Path


class Storage(ABC):
    @abstractmethod
    def get(self, key: str) -> bytes: ...

    @abstractmethod
    def put(self, key: str, data: bytes) -> None: ...

    @abstractmethod
    def list_prefix(self, prefix: str) -> list[str]: ...

    @abstractmethod
    def uri(self, key: str) -> str: ...


class LocalStorage(Storage):
    def __init__(self, root: str | Path) -> None:
        self.root = Path(root).resolve()

    def _resolve(self, key: str) -> Path:
        path = (self.root / key).resolve()
        if not path.is_relative_to(self.root):
            raise ValueError(f"key escapes storage root: {key}")
        return path

    def get(self, key: str) -> bytes:
        return self._resolve(key).read_bytes()

    def put(self, key: str, data: bytes) -> None:
        path = self._resolve(key)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(data)

    def list_prefix(self, prefix: str) -> list[str]:
        base = self._resolve(prefix)
        if not base.is_dir():
            return []
        return sorted(
            str(p.relative_to(self.root))
            for p in base.rglob("*")
            if p.is_file()
        )

    def uri(self, key: str) -> str:
        return f"file://{self._resolve(key)}"


class GcsStorage(Storage):
    def __init__(self, bucket_name: str) -> None:
        from google.cloud import storage as gcs

        self.bucket_name = bucket_name
        self._bucket = gcs.Client().bucket(bucket_name)

    def get(self, key: str) -> bytes:
        return self._bucket.blob(key).download_as_bytes()

    def put(self, key: str, data: bytes) -> None:
        self._bucket.blob(key).upload_from_string(data)

    def list_prefix(self, prefix: str) -> list[str]:
        return sorted(b.name for b in self._bucket.list_blobs(prefix=prefix))

    def uri(self, key: str) -> str:
        return f"gs://{self.bucket_name}/{key}"
